src/Workspaces/Remote/Core/BrokeredServiceConnection.cs

```diff
@@ -161,11 +162,11 @@ public override async ValueTask<Optional<TResult>> TryInvokeAsync<TResult>(Solut
                 {
                     await invocation(service, pipe.Writer, cancellationToken).ConfigureAwait(false);
                 }
-                catch (Exception e)
+                catch (Exception e) when (e is not (OperationCanceledException or RemoteInvocationException))
```

Comment:
📝 I'm not sure this is actually correct. `pipe.Writer` in this scenario is an _argument_ to the call, as opposed to the pipe actually being used for the message (i.e. there are two writers involved, and `pipe.Writer` would not be the one automatically closed in these cases).

Reply:
@AArnott

Reply:
My understanding is that if the exception comes from the proxy, the proxy will take care of completing the writer argument. If it doesn't come from proxy (it's from the code surrounding the call to the proxy in `invocation` delegate) then we need to complete the writer argument. The only exceptions that we expect to come from the proxy are `OperationCanceledException` and `RemoteInvocationException`. Maybe should also add `MessagePackSerializationException`.

Reply:
"The proxy"  isn't doing any of this. It's all JsonRpc which the proxy is driving. I imagine that's what you had in mind, but just in case I wanted to clarify.
OCE might be thrown by JsonRpc *before* even processing the arguments or it might come from the server when it responded to cancellation. In the former case, the `Pipe.Writer` will not have been completed and in the latter case it will. So in some cases it's undefined whether the Pipe.Writer is always completed.
Now, completing an already completed Writer is harmless. You just had better not do it *concurrently* with use or completion of the Writer. So completing the Writer in your catch block without the exception filter *should* be safe AFAIK.

Reply:
OK, let's just leave it as is then.